pkg/reconciler/labeler/v2/accessors.go

```diff
@@ -57,9 +59,37 @@ func NewRevisionAccessor(
 	}
 }
 
-// get implements Accessor
-func (r *Revision) get(ns, name string) (kmeta.Accessor, error) {
-	return r.revisionLister.Revisions(ns).Get(name)
+// makeMetadataPatch makes a metadata map to be patched or nil if no changes are needed.
+func makeMetadataPatch(acc kmeta.Accessor, routeName *string) (map[string]interface{}, error) {
+	labels := map[string]interface{}{}
+
+	if err := addRouteLabel(acc, labels, routeName); err != nil {
+		return nil, err
+	}
+
+	if len(labels) > 0 {
+		meta := map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels": labels,
+			},
+		}
+		return meta, nil
+	}
+	return nil, nil
+}
+
+// addRouteLabel appends the route label to the list of labels if needed.
+func addRouteLabel(acc kmeta.Accessor, labels map[string]interface{}, routeName *string) error {
+	if oldLabels := acc.GetLabels(); oldLabels == nil && routeName != nil {
+		labels[serving.RouteLabelKey] = routeName
+	} else if oldLabel := oldLabels[serving.RouteLabelKey]; routeName == nil && oldLabel != "" {
+		labels[serving.RouteLabelKey] = routeName
```

Comment:
This is probably incorrect.

Reply:
I've expanded it out a bit which I think is easier to follow?